python/services/dataservice/dmod/dataservice/service.py

```diff
@@ -643,6 +662,59 @@ def _process_query(self, message: DatasetManagementMessage) -> DatasetManagement
             reason = 'Unsupported {} Query Type - {}'.format(DatasetQuery.__class__.__name__, query_type.name)
             return DatasetManagementResponse(action=message.management_action, success=False, reason=reason)
 
+    def _temp_dataset_mark_and_sweep(self):
+        """
+        Encapsulation of the required behavior for a single iteration through ::method:`manage_temp_datasets`.
+
+        Method scans for in use temporary datasets, updating the expire times of each.  It then deletes any expired
+        datasets marked for deletion already within the instance's private field.  Finally, it updates the instance's
+        private field for marked-for-deletion datasets with any that are now expired (and thus will be removed on the
+        next invocation).
+        """
+        # Get these upfront, since no meaningful changes can happen during the course of an iteration
+        temp_datasets = {ds_name: ds for ds_name, ds in self.get_known_datasets().values() if ds.is_temporary}
```

Comment:
I think this is fine so long as temp datasets can only be used by a single job. If a temp dataset can be referenced by multiple jobs, I think there is a data race if a job starts that uses a temp dataset but the dataservice is waiting to check for new jobs. The temp dataset that job depends on could be swept.

Reply:
Good catch.   I'll need to put something in `ServiceManager.perform_checks_for_job` to block the mark and sweep routine (or maybe even just the sweep portion).  I'll also either have to unmark in this function, or perhaps move the logic for linking job users to here.

Reply:
I have made some changes to account for this.